### PR TITLE
Factor TokenTrees into their own file in preparation for TokenStreams.

### PR DESCRIPTION
**Pull request by cgswords**
syntax-[breaking-change] cc #31645 
(Only breaking because ast::TokenTree is now tokenstream::TokenTree.)

This pull request refactors TokenTrees into their own file as src/libsyntax/tokenstream.rs, moving them out of src/libsyntax/ast.rs, in order to prepare for an accompanying TokenStream implementation (per RFC 1566).

r? @nrc 

(Note: duplicate pull request because I destroyed the previous branch last time on accident.)

### CODE REVIEW

**Comment by nrc** [comment]
cc @Manishearth (breaking batch)

**Comment by nrc** [comment]
r=me

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -19,13 +19,10 @@ use attr::{ThinAttributes, HasAttrs};
...
 use abi::Abi;
 use errors;
-use ext::base;
-use ext::tt::macro_parser;
 use parse::token::{self, keywords, InternedString};
-use parse::lexer;
-use parse::lexer::comments::{doc_comment_style, strip_doc_comment_decoration};
 use print::pprust;
 use ptr::P;
+use tokenstream::{TokenTree};
```
Comment (durka) [inline src/libsyntax/ast.rs]: I'll make the same comment as last time: is it worth changing this `use` to a `pub use` in order to avoid the breakage?

Comment (cgswords) [inline src/libsyntax/ast.rs]: IMO, no. Token streams will eventually be added to `tokenstream.rs`, and hopefully places that currently use tokens trees will generally move to using token streams. Looking forward, procedural macros will shift to work over token streams, and the breakage here is (at least partially) intended to  ensure that tokentree actions / procedural macros no longer need to rely on `ast.rs` in any way.

**Comment by Manishearth** [comment]
travis fails

**Comment by cgswords** [comment]
Fixed the travis failure.

**Pull request merged**